pkg/ruler/rule_concurrency.go

```diff
@@ -0,0 +1,213 @@
+	return &MultiTenantConcurrencyControllerMetrics{
+		CurrentConcurrency: promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+			Name: "cortex_ruler_global_rule_evaluation_concurrency_current",
+			Help: "Current number of active global concurrency slots",
+		}),
+		AcquireTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_ruler_global_rule_evaluation_concurrency_acquire_total",
+			Help: "Total number of acquired concurrency slots",
+		}),
+		AcquireFailedTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_ruler_global_rule_evaluation_concurrency_acquire_failed_total",
+			Help: "Total number of failed attempts to acquire concurrency slots",
+		}),
+		ReleaseTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_ruler_global_rule_evaluation_concurrency_release_total",
+			Help: "Total number of released concurrency slots",
+		}),
+	}
+}
+
+// MultiTenantConcurrencyController is a concurrency controller that limits the number of concurrent rule evaluations both global and per tenant.
+type MultiTenantConcurrencyController struct {
+	logger               log.Logger
+	limits               RulesLimits
+	maxGlobalConcurrency int64
+	metrics              *MultiTenantConcurrencyControllerMetrics
+
+	globalConcurrency semaphore.Weighted
+	mtx               sync.Mutex
+	tenantConcurrency map[string]*DynamicSemaphore
+}
+
+// NewMultiTenantConcurrencyController creates a new MultiTenantConcurrencyController.
+func NewMultiTenantConcurrencyController(logger log.Logger, maxGlobalConcurrency int64, limits RulesLimits, reg prometheus.Registerer) *MultiTenantConcurrencyController {
+	return &MultiTenantConcurrencyController{
+		metrics:              newMultiTenantConcurrencyControllerMetrics(reg),
+		logger:               log.With(logger, "component", "ruler.MultiTenantConcurrencyController"),
+		maxGlobalConcurrency: maxGlobalConcurrency,
+		limits:               limits,
+		globalConcurrency:    *semaphore.NewWeighted(maxGlobalConcurrency),
+		tenantConcurrency:    make(map[string]*DynamicSemaphore),
+	}
+}
+
+// Done releases a slot from the concurrency controller.
+func (c *MultiTenantConcurrencyController) Done(ctx context.Context) {
+	c.globalConcurrency.Release(1)
+	c.metrics.CurrentConcurrency.Dec()
+	c.metrics.ReleaseTotal.Inc()
+
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		level.Error(c.logger).Log("msg", "error extracting org id from context while releasing a slot", "err", err)
+		return
+	}
+
+	c.mtx.Lock()
+	tc, ok := c.tenantConcurrency[userID]
+	c.mtx.Unlock()
+
+	if ok {
+		tc.Release()
+	} else {
+		c.logger.Log("msg", "tenant concurrency controller: tenant not found", "tenant", userID)
+	}
+}
+
+// Allow tries to acquire a slot from the concurrency controller.
+func (c *MultiTenantConcurrencyController) Allow(ctx context.Context, group *rules.Group, rule rules.Rule) bool {
+	c.metrics.AcquireTotal.Inc()
```

Comment:
The metric doc says "Total number of acquired concurrency slots" but it's not true. We should increase the metric only when we return true from this function.

Reply:
We discussed this offline, this is actually correct. I want to compute successes as `total - failed` as such, I need acquire to be increment regarless of what happens. I have adjusted the metric names and description in:

https://github.com/grafana/mimir/pull/8146/commits/004f3a11b2208bc5318a49460b1e10fc89d7e3bd

Let me know what you think.